cmd/benchttp/run.go

```diff
@@ -0,0 +1,151 @@
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+
+	"github.com/benchttp/runner/config"
+	"github.com/benchttp/runner/internal/configfile"
+	"github.com/benchttp/runner/internal/configflags"
+	"github.com/benchttp/runner/output"
+	"github.com/benchttp/runner/requester"
+)
+
+var defaultConfigFiles = []string{
+	"./.benchttp.yml",
+	"./.benchttp.yaml",
+	"./.benchttp.json",
+}
+
+// cmdRun handles subcommand "benchttp run [options]".
+type cmdRun struct {
+	flagset *flag.FlagSet
+
+	// defaultConfigFiles is a slice of default config files to look up for
+	// by order of priority if none is provided via the -configFile flag.
+	defaultConfigFiles []string
+
+	// configFile is the parsed value for flag -configFile
+	configFile string
+
+	// config is the runner config resulting from parsing CLI flags.
+	config config.Global
+}
+
+// ensure cmdRun implements command
+var _ command = (*cmdRun)(nil)
+
+// execute runs the benchttp runner: it parses CLI flags, loads config
+// from config file and parsed flags, then runs the benchmark and outputs
+// it according to the config.
+func (cmd cmdRun) execute(args []string) error {
+	fieldsSet := cmd.parseArgs(args)
+
+	cfg, err := cmd.makeConfig(fieldsSet)
+	if err != nil {
+		return err
+	}
+
+	req, err := cfg.Request.Value()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go cmd.listenOSInterrupt(cancel)
+
+	rep, err := requester.New(cmd.requesterConfig(cfg)).Run(ctx, req)
+	if err != nil {
+		if errors.Is(err, requester.ErrCanceled) {
+			if err := cmd.handleRunInterrupt(); err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+
+	return output.New(rep, cfg).Export()
+}
+
+// parseArgs parses input args as config fields and returns
+// a slice of fields that were set by the user.
+func (cmd *cmdRun) parseArgs(args []string) []string {
+	// first arg is subcommand "run"
+	if len(args) <= 1 {
+		return []string{}
+	}
+
+	// config file path
+	cmd.flagset.StringVar(&cmd.configFile,
+		"configFile",
+		configfile.Find(cmd.defaultConfigFiles),
+		"Config file path",
+	)
+
+	// cli config
+	configflags.Set(cmd.flagset, &cmd.config)
+
+	cmd.flagset.Parse(args[1:]) //nolint:errcheck // never occurs due to flag.ExitOnError
+
+	return configflags.Which(cmd.flagset)
+}
+
+// makeConfig returns a config.Config initialized with config file
+// options if found, overridden with CLI options listed in fields
+// slice param.
+func (cmd cmdRun) makeConfig(fields []string) (cfg config.Global, err error) {
+	fileConfig, err := configfile.Parse(cmd.configFile)
+	if err != nil && !errors.Is(err, configfile.ErrFileNotFound) {
+		// config file is not mandatory, other errors are critical
+		return
+	}
+
+	mergedConfig := fileConfig.Override(cmd.config, fields...)
+
+	return mergedConfig, mergedConfig.Validate()
+}
+
+// requesterConfig returns a requester.Config generated from cfg.
+func (cmd cmdRun) requesterConfig(cfg config.Global) requester.Config {
```

Comment:
Nitpick: it seems it does not need to be a method of `cmdRun`.

I understand it is to limit the access to this to `cmdRun`, but is it a good enough reason ? Really question, I'd like you're view on that.

Reply:
Well 2 reasons:
- namespacing indeed, to avoid polluting the package scope that is expected to manage very different things
- its i/o `config.Global` and `requester.Config` are 100% specific to the `run` command (they have nothing to do with with other commands)

So to me it makes perfect sense as a method of `cmdRun`. Could also be a function of some package `internal/adapters` but we stated it's bit overkill for now. I don't mind going this way though, otherwise I'd rather keep the package as clean as possible.

Side question: what do you mean exactly by _it does not need_? It seems to imply that it comes at some sort of cost compared to package-scoped functions, what would that be?
If it's only the extra copy of `cmdRun` we can still use a pointer receiver instead. But note that the structure is 208 bytes (we set the linter `hugeParam` to complain at 256) and this method is not expected to be called more than once in the whole program, so I think we're fine with that.

Reply:
> Side question: what do you mean exactly by it does not need?

Because it really does not _need_ the method receiver `(cmd cmdRun)`.
Namespacing this way is valid too. I'm totally fine with doing this.